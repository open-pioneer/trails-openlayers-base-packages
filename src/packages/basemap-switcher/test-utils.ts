// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
