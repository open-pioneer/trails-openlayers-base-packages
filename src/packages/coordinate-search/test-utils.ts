// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
