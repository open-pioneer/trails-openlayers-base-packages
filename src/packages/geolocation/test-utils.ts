// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
