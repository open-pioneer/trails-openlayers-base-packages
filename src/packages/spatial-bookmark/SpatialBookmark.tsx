// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    Box,
    Flex,
    List,
    ListItem,
    ListIcon,
    Text,
    Button
} from "@open-pioneer/chakra-integration";
import { PackageIntl } from "@open-pioneer/runtime";
import { useMapModel } from "@open-pioneer/map";
import { useIntl } from "open-pioneer:react-hooks";
import { AddIcon } from "@chakra-ui/icons";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { PiMapTrifold, PiTrashSimpleLight } from "react-icons/pi";
import { FC, useEffect, useState } from "react";
const INIT_BOOKMARKS = [
    { title: "Düsseldorf", view: {} },
    { title: "Langenfeld", view: {} },
    { title: "Köln", view: {} }
];
interface SpatialBookmarkProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;
}
export const SpatialBookmark: FC<SpatialBookmarkProps> = (props) => {
    const { mapId } = props;
    const intl = useIntl();
    const [bookmarks, setBookmarks] = useState(INIT_BOOKMARKS);
    const { containerProps } = useCommonComponentProps("spatial-bookmark", props);
    const bookmarkList = createList(bookmarks);

    return (
        <Box {...containerProps} backgroundColor="white" padding={2} width={"350px"}>
            {bookmarkList}
            <BookmarkHandler intl={intl} />
        </Box>
    );
};
function createList(bookmarks: { title: string; view: {} }[]) {
    const bookmarkItems = bookmarks.map((bookmark, idx) => (
        <BookmarkItem key={idx} title={bookmark.title} />
    ));

    return (
        <List as="ul" className="spatial-bookmark-list" listStyleType="none" spacing={3}>
            {bookmarkItems}
        </List>
    );
}
function BookmarkItem(props: { title: string }): JSX.Element {
    const { title } = props;
    return (
        <ListItem>
            <Text noOfLines={1}>
                <ListIcon as={PiMapTrifold} />
                {title}
            </Text>
        </ListItem>
    );
}
function BookmarkHandler(props: { intl: PackageIntl }): JSX.Element {
    const { intl } = props;
    return (
        <Flex width="100%" flexDirection="row" gap={1} my={2}>
            <Button
                size="sm"
                variant="outline"
                iconSpacing={0}
                leftIcon={<PiTrashSimpleLight />}
            ></Button>
            <Button size="sm" variant="outline" leftIcon={<AddIcon />}>
                {intl.formatMessage({ id: "createBookmarkLabel" })}
            </Button>
        </Flex>
    );
}
