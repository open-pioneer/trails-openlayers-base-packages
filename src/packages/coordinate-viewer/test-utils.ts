// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
