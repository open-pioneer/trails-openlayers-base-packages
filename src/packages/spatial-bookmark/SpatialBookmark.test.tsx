// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { afterEach, expect, it } from "vitest";

import {
    render,
    screen,
    findByRole,
    findByPlaceholderText,
    findByText,
    findByLabelText
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import { LocalStorageService, LocalStorageNamespace } from "@open-pioneer/local-storage";
import { SpatialBookmark } from "./SpatialBookmark";
import { Bookmark } from "./SpatialBookmarkViewModel";

/*
    TODO: 
    - replace toMatchInlineSnapshot with toMatchSnapshot ?!
*/

const MOCK_NAMESPACE = new Map<string, unknown>();
afterEach(() => {
    MOCK_NAMESPACE.clear();
});
const localStorageService: Partial<LocalStorageService> = {
    getNamespace(key): LocalStorageNamespace {
        if (key !== "spatial-bookmarks") {
            throw new Error("unexpected key");
        }
        return {
            get(key) {
                return MOCK_NAMESPACE.get(key);
            },
            set(key, value) {
                MOCK_NAMESPACE.set(key, value);
            }
        } as LocalStorageNamespace;
    }
};
const packageNamespace = localStorageService.getNamespace!("spatial-bookmarks");

it("except if currently no bookmarks are saved to display info message", async () => {
    packageNamespace.set("bookmarks", []);
    createBookmarkComponent();
    const div = await waitForSpatialBookmark();
    const alertDiv = await findByRole(div, "alert");
    expect(alertDiv).toMatchInlineSnapshot(`
      <div
        class="chakra-alert css-b9ls1z"
        data-status="info"
        data-theme="light"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-14ogjxt"
          data-status="info"
          data-theme="light"
        >
          <svg
            class="chakra-icon css-qk6lof"
            data-theme="light"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
              fill="currentColor"
            />
          </svg>
        </span>
        bookmark.alert.noSaved
      </div>
    `);
});
it("should load the stored bookmark from localstorage and present it visibly within the list", async () => {
    createBookmark();
    createBookmarkComponent();
    const div = await waitForSpatialBookmark();
    const bookmarkList = await findByRole(div, "listbox");
    expect(bookmarkList).toMatchInlineSnapshot(`
      <ul
        aria-label="bookmark.list.label"
        class="spatial-bookmark-list css-y3cje8"
        data-theme="light"
        role="listbox"
      >
        <li
          class="spatial-bookmark-item spatial-bookmark-item-test css-ol9amj"
          data-theme="light"
          role="option"
        >
          <div
            class="css-1h36yqy"
            data-theme="light"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M228.92,49.69a8,8,0,0,0-6.86-1.45L160.93,63.52,99.58,32.84a8,8,0,0,0-5.52-.6l-64,16A8,8,0,0,0,24,56V200a8,8,0,0,0,9.94,7.76l61.13-15.28,61.35,30.68A8.15,8.15,0,0,0,160,224a8,8,0,0,0,1.94-.24l64-16A8,8,0,0,0,232,200V56A8,8,0,0,0,228.92,49.69ZM104,52.94l48,24V203.06l-48-24ZM40,62.25l48-12v127.5l-48,12Zm176,131.5-48,12V78.25l48-12Z"
              />
            </svg>
            <p
              class="chakra-text css-zvlevn"
              data-theme="light"
            >
              Test
            </p>
            <div
              class="css-17xejub"
              data-theme="light"
            />
            <button
              aria-label="bookmark.button.deleteOne"
              class="chakra-button spatial-bookmark-item-details css-hzhyv0"
              data-theme="light"
              type="button"
            >
              <span
                class="chakra-button__icon css-1dwr89y"
                data-theme="light"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M216,50H40a6,6,0,0,0,0,12H50V208a14,14,0,0,0,14,14H192a14,14,0,0,0,14-14V62h10a6,6,0,0,0,0-12ZM194,208a2,2,0,0,1-2,2H64a2,2,0,0,1-2-2V62H194ZM82,24a6,6,0,0,1,6-6h80a6,6,0,0,1,0,12H88A6,6,0,0,1,82,24Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </li>
      </ul>
    `);
});

it("should save a bookmark and update the list", async () => {
    const user = userEvent.setup();
    packageNamespace.set("bookmarks", []);
    createBookmarkComponent();
    const div = await waitForSpatialBookmark();
    const createBtn = await findByText(div, "bookmark.button.create");

    //click the button to create a bookmark
    await user.click(createBtn);
    const input = await findByPlaceholderText(div, "bookmark.input.placeholder");
    const saveBtn = await findByText(div, "bookmark.button.save");
    //set bookmark name in the input
    await user.type(input, "Dortmund");
    await user.click(saveBtn);
    const bookmark = packageNamespace.get("bookmarks") as Bookmark[];
    //TODO: snapshot differs due to the slug method in listItem css class
    expect(bookmark[0]?.title).toEqual("Dortmund");
});

it("should only delete all bookmarks after the user confirms the action.", async () => {
    createBookmark();
    const user = userEvent.setup();
    createBookmarkComponent();
    const div = await waitForSpatialBookmark();
    const deleteAllBtn = await findByLabelText(div, "bookmark.button.deleteAll");
    await user.click(deleteAllBtn);
    const alertMessage = await findByRole(div, "alert");
    expect(alertMessage).toMatchInlineSnapshot(`
      <div
        class="chakra-alert css-ercdzk"
        data-status="warning"
        data-theme="light"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-14ogjxt"
          data-status="warning"
          data-theme="light"
        >
          <svg
            class="chakra-icon css-qk6lof"
            data-theme="light"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
              fill="currentColor"
            />
          </svg>
        </span>
        bookmark.alert.delete
      </div>
    `);
    const confirmBtn = await findByText(div, "bookmark.button.confirmDelete");
    await user.click(confirmBtn);
    const noSavedBookmarksAlert = await findByText(div, "bookmark.alert.noSaved");
    expect(noSavedBookmarksAlert).toMatchInlineSnapshot(`
      <div
        class="chakra-alert css-b9ls1z"
        data-status="info"
        data-theme="light"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-14ogjxt"
          data-status="info"
          data-theme="light"
        >
          <svg
            class="chakra-icon css-qk6lof"
            data-theme="light"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
              fill="currentColor"
            />
          </svg>
        </span>
        bookmark.alert.noSaved
      </div>
    `);
});

it("should remove bookmark from the list by clicking the remove button", async () => {
    createBookmark();
    const user = userEvent.setup();
    createBookmarkComponent();
    const div = await waitForSpatialBookmark();
    const deleteBookmarkBtn = await findByLabelText(div, "bookmark.button.deleteOne");
    await user.click(deleteBookmarkBtn);
    const noSavedBookmarksAlert = await findByText(div, "bookmark.alert.noSaved");
    expect(noSavedBookmarksAlert).toMatchInlineSnapshot(`
      <div
        class="chakra-alert css-b9ls1z"
        data-status="info"
        data-theme="light"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-14ogjxt"
          data-status="info"
          data-theme="light"
        >
          <svg
            class="chakra-icon css-qk6lof"
            data-theme="light"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
              fill="currentColor"
            />
          </svg>
        </span>
        bookmark.alert.noSaved
      </div>
    `);
});

it("should center the map on the selected bookmark", async () => {
    const bookmark = createBookmark();
    const user = userEvent.setup();
    const { mapId, injectedServices, mapModel } = await setupComponent();
    renderComponent({ services: injectedServices }, { mapId: mapId, "data-testid": "bookmarks" });
    const extent = bookmark.extent;
    const bookmarkExtent = [extent.minX, extent.minY, extent.maxX, extent.maxY];
    const initExtent = mapModel.olMap.getView().calculateExtent();
    expect(initExtent).not.toStrictEqual(bookmarkExtent);
    const div = await waitForSpatialBookmark();
    const bookmarkItem = await findByText(div, "Test");
    await user.click(bookmarkItem);
    const newExtent = mapModel.olMap.getView().calculateExtent();
    expect(newExtent).toStrictEqual(bookmarkExtent);
});

interface ctxProviderConfig {
    "local-storage.LocalStorageService": Partial<LocalStorageService>;
}
interface componentPropsConfig {
    mapId: string;
    "data-testid": string;
}
async function createBookmarkComponent() {
    const { mapId, injectedServices } = await setupComponent();
    renderComponent({ services: injectedServices }, { mapId: mapId, "data-testid": "bookmarks" });
}
async function setupComponent() {
    const { mapId, registry } = await setupMap({
        projection: "EPSG:25832"
    });
    const mapModel = await registry.expectMapModel(mapId);
    const injectedServices = {
        ...createServiceOptions({ registry }),
        "local-storage.LocalStorageService": localStorageService
    };
    return {
        mapModel,
        mapId,
        injectedServices
    };
}
function renderComponent(
    contextProviderProps: { services: ctxProviderConfig },
    componentProps: componentPropsConfig
) {
    return render(
        <PackageContextProvider {...contextProviderProps}>
            <SpatialBookmark {...componentProps} />
        </PackageContextProvider>
    );
}
async function waitForSpatialBookmark() {
    return await screen.findByTestId<HTMLDivElement>("bookmarks");
}

function createBookmark() {
    const bookmark: Bookmark = {
        id: "test",
        title: "Test",
        extent: {
            minX: 404609.0492711461,
            minY: 5759938.978042119,
            maxX: 405066.2854485374,
            maxY: 5760396.21421951
        },
        projection: "EPSG:25832"
    };
    packageNamespace.set("bookmarks", [bookmark]);
    return bookmark;
}
