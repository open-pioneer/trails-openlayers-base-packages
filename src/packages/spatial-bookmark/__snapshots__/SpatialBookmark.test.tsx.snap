// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should load the stored bookmark from local storage and present it visibly within the list 1`] = `
<ul
  aria-label="bookmark.list.label"
  class="spatial-bookmark-list css-2y973l"
  data-theme="light"
  role="listbox"
>
  <li
    class="spatial-bookmark-item css-1npgjds"
    data-theme="light"
    role="option"
    tabindex="0"
  >
    <div
      class="css-1h36yqy"
      data-theme="light"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M228.92,49.69a8,8,0,0,0-6.86-1.45L160.93,63.52,99.58,32.84a8,8,0,0,0-5.52-.6l-64,16A8,8,0,0,0,24,56V200a8,8,0,0,0,9.94,7.76l61.13-15.28,61.35,30.68A8.15,8.15,0,0,0,160,224a8,8,0,0,0,1.94-.24l64-16A8,8,0,0,0,232,200V56A8,8,0,0,0,228.92,49.69ZM104,52.94l48,24V203.06l-48-24ZM40,62.25l48-12v127.5l-48,12Zm176,131.5-48,12V78.25l48-12Z"
        />
      </svg>
      <p
        class="chakra-text css-1uvjhra"
        data-theme="light"
      >
        Test
      </p>
      <div
        class="css-17xejub"
        data-theme="light"
      />
      <button
        aria-label="bookmark.button.deleteOne"
        class="chakra-button spatial-bookmark-item-delete css-kxp9k4"
        data-theme="light"
        type="button"
      >
        <span
          class="chakra-button__icon css-1dwr89y"
          data-theme="light"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M216,50H40a6,6,0,0,0,0,12H50V208a14,14,0,0,0,14,14H192a14,14,0,0,0,14-14V62h10a6,6,0,0,0,0-12ZM194,208a2,2,0,0,1-2,2H64a2,2,0,0,1-2-2V62H194ZM82,24a6,6,0,0,1,6-6h80a6,6,0,0,1,0,12H88A6,6,0,0,1,82,24Z"
            />
          </svg>
        </span>
      </button>
    </div>
  </li>
</ul>
`;

exports[`should only delete all bookmarks after the user confirms the action. 1`] = `
<div
  class="chakra-alert css-134lewa"
  data-status="warning"
  data-theme="light"
  role="alert"
>
  <span
    class="chakra-alert__icon css-14ogjxt"
    data-status="warning"
    data-theme="light"
  >
    <svg
      class="chakra-icon css-qk6lof"
      data-theme="light"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
        fill="currentColor"
      />
    </svg>
  </span>
  bookmark.alert.delete
</div>
`;

exports[`should only delete all bookmarks after the user confirms the action. 2`] = `
<div
  class="chakra-alert css-1sqgcxg"
  data-status="info"
  data-theme="light"
  role="alert"
>
  <span
    class="chakra-alert__icon css-14ogjxt"
    data-status="info"
    data-theme="light"
  >
    <svg
      class="chakra-icon css-qk6lof"
      data-theme="light"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
        fill="currentColor"
      />
    </svg>
  </span>
  bookmark.alert.noSaved
</div>
`;

exports[`should remove bookmark from the list by clicking the remove button 1`] = `
<div
  class="chakra-alert css-1sqgcxg"
  data-status="info"
  data-theme="light"
  role="alert"
>
  <span
    class="chakra-alert__icon css-14ogjxt"
    data-status="info"
    data-theme="light"
  >
    <svg
      class="chakra-icon css-qk6lof"
      data-theme="light"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
        fill="currentColor"
      />
    </svg>
  </span>
  bookmark.alert.noSaved
</div>
`;

exports[`should shown an alert message if no bookmarks are saved 1`] = `
<div
  class="chakra-alert css-1sqgcxg"
  data-status="info"
  data-theme="light"
  role="alert"
>
  <span
    class="chakra-alert__icon css-14ogjxt"
    data-status="info"
    data-theme="light"
  >
    <svg
      class="chakra-icon css-qk6lof"
      data-theme="light"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
        fill="currentColor"
      />
    </svg>
  </span>
  bookmark.alert.noSaved
</div>
`;
